cosmpy/aerial/contract/__init__.py

```
@@ -0,0 +1,263 @@
+
+import json
+from datetime import datetime
+from typing import Any, Optional
+
+from cosmpy.aerial.client import LedgerClient
+from cosmpy.aerial.contract.cosmwasm import (
+    create_cosmwasm_execute_msg,
+    create_cosmwasm_instantiate_msg,
+    create_cosmwasm_store_code_msg,
+)
+from cosmpy.aerial.tx import SigningCfg, Transaction
+from cosmpy.aerial.tx_helpers import SubmittedTx
+from cosmpy.aerial.wallet import Wallet
+from cosmpy.crypto.address import Address
+from cosmpy.crypto.hashfuncs import sha256
+from cosmpy.protos.cosmos.base.query.v1beta1.pagination_pb2 import PageRequest
+from cosmpy.protos.cosmwasm.wasm.v1.query_pb2 import (
+    QueryCodesRequest,
+    QuerySmartContractStateRequest,
+)
+
+
+def _compute_digest(path: str) -> bytes:
+    with open(path, "rb") as input_file:
+        return sha256(input_file.read())
+
+
+def _generate_label(digest: bytes) -> str:
+    now = datetime.utcnow()
+    return f"{digest.hex()[:14]}-{now.strftime('%Y%m%d%H%M%S')}"
+
+
+class LedgerContract:
+    def __init__(
+        self, path: str, client: LedgerClient, address: Optional[Address] = None
+    ):
+        self._path = path
+        self._client = client
+        self._digest = _compute_digest(self._path)
+        self._code_id = self._find_contract_id_by_digest(self._digest)
+        self._address = address
+
+    @property
+    def path(self) -> str:
+        return self._path
+
+    @property
+    def digest(self) -> bytes:
+        return self._digest
+
+    @property
+    def code_id(self) -> Optional[int]:
+        return self._code_id
+
+    @property
+    def address(self) -> Optional[Address]:
+        return self._address
+
+    def store(self, sender: Wallet, gas_limit: Optional[int] = None) -> int:
+
+        # query the account information for the sender
+        account = self._client.query_account(sender.address())
+
+        # estimate the fee required for this transaction
+        gas_limit = (
+            gas_limit or 2000000
```

In future PRs I think we should query the maximum gas limit from node and clamp gas limit if is more than max.

In case of gas_limit higher than max you get silent error - tx never gets on chain without giving any feedback.

---

Agreed - updated the description of #100